gpflow/misc.py

```diff
@@ -251,5 +251,57 @@ def _get_tensor_safe(name, index, graph):
     except KeyError:
         return None
 
+
+def swap_final_dims(X: tf.Tensor):
+    """
+    equivalent to np.swapaxes(X, -2, -1)
+    """
+    K = tf.rank(X)
+    num_broadcasting_dims = tf.reduce_max([K - 2, 0])
+    perm = tf.concat([tf.range(num_broadcasting_dims), [K - 1, K - 2]], 0)
+    return tf.transpose(X, perm)
+
+
+def slice_final_dim(X: tf.Tensor, sl: slice, dim: int):
+    """
+    Takes the slice sl from the final dimension of X
+    If the slice is the whole of the final dimension (i.e. sl=slice(0, dim)) then there is nothing to do
+    NB tensorflow does not support slicing with step!=1, so raise error if this is attempted
+    """
+    if (sl.start is None) and (sl.step is None) and (sl.stop == dim):  # nothing to do
+        return X
+    else:
+        if sl.step and (sl.step != 1):
+            raise NotImplementedError('only slices with step=1 supported by tensorflow')
+
+        start = 1 if sl.start is None else sl.start
+
+        begin = tf.concat([tf.zeros([tf.rank(X) - 1], dtype=settings.int_type),
+                           tf.reshape(start, [1])], 0)
+
+        size = tf.concat([tf.shape(X)[:-1], tf.reshape(sl.stop - start, [1])], 0)
+
+        return tf.slice(X, begin, size)
+
+
+def tile_over_new_axis(X : tf.Tensor, repeats : tf.Tensor, axis : int):
+    """
+    Inserts a new axis and tiles k times
+    e.g. if X.shape = [S, N, D], and axis=-2 then this is equivalent to
+     tf.tile(X[:, :, None, :], [1, 1, repeats, 1])
```

Comment:
* Formatting issues - space before "tf.tile" and "supports negative ..."
* Would be nice to have docstrings for parameters